parsec/parsec_mpi_funnelled.c

```diff
@@ -603,108 +610,173 @@ mpi_funnelled_fini(parsec_comm_engine_t *ce)
+    assert(MPI_COMM_NULL == parsec_ce_mpi_am_comm[0] );  /* no communicator */
     MAX_MPI_TAG = -1;  /* mark the layer as uninitialized */
     size_of_total_reqs = 0;
     mpi_funnelled_last_active_req = 0;
     mpi_funnelled_static_req_idx = 0;
 
-    nb_internal_tag = 0;
-    count_internal_tag = 0;
-
     return 1;
 }
 
-/* Users need to register all tags before finalizing the comm
- * engine init.
- * The requested tags should be from 0 up to MPI_FUNNELLED_MAX_TAG,
- * dynamic tags will start from MPI_FUNNELLED_MAX_TAG.
+/**
+ * @brief Register AM tags. Classical API, a tag a callback function and a callback data
+ *        for the message arrival and a maximum length of the AM message for preallocating
+ *        all necessary buffers. The registration can happen at any moment, before or after
+ *        the communication engine was started. If the communication engine is inactive,
+ *        the tags will be stored but anything related to the communication registration of
+ *        such AM will be delayed until the engine starts. If the engine is running, we
+ *        notify it that the AM infrastructure needs to be rebuilt, and it will do it at
+ *        the next progress cycle.
  */
 int
 mpi_no_thread_tag_register(parsec_ce_tag_t tag,
                            parsec_ce_am_callback_t callback,
                            void *cb_data,
                            size_t msg_length)
 {
-    mpi_funnelled_callback_t *cb;
-
     /* All internal tags have been registered */
-    if(nb_internal_tag == count_internal_tag) {
-        if(tag >= PARSEC_MAX_REGISTERED_TAGS) {
-            parsec_warning("Tag is out of range, it has to be between %d - %d\n", 0, PARSEC_MAX_REGISTERED_TAGS);
-            return PARSEC_ERR_VALUE_OUT_OF_BOUNDS;
-        }
+    if(tag >= PARSEC_MAX_REGISTERED_TAGS) {
+        parsec_warning("Tag is out of range, it has to be between %d - %d\n", 0, PARSEC_MAX_REGISTERED_TAGS);
+        return PARSEC_ERR_VALUE_OUT_OF_BOUNDS;
     }
     mpi_funnelled_tag_t *tag_struct = &parsec_mpi_funnelled_array_of_registered_tags[tag];
-    if(NULL != parsec_mpi_funnelled_array_of_registered_tags[tag].buf ) {
-        parsec_warning("Tag: %d is already registered\n", (int)tag);
+    if(NULL != parsec_mpi_funnelled_array_of_registered_tags[tag].callback ) {
+        parsec_warning("Tag: %d is already registered (callback %p, callback data %p, msg length %d)\n",
+                       (int)tag, tag_struct->callback, tag_struct->cb_data, (int)tag_struct->msg_length);
         return PARSEC_ERR_EXISTS;
     }
+    parsec_atomic_lock(&parsec_ce_am_build_lock);
+    tag_struct->tag = tag;
+    tag_struct->msg_length = (msg_length + 15) & ~0xF;  /* align to 16 bytes */
+    tag_struct->callback = callback;
+    tag_struct->cb_data = cb_data;
+    tag_struct->status = PARSEC_CE_TAG_STATUS_ENABLE;
 
+    /* Update the total number of requests we know about */
     size_of_total_reqs += EACH_STATIC_REQ_RANGE;
```

Comment:
Shouldn't this be done in `parsec_ce_rebuild_am_requests`? I'm worried that otherwise the communication will see `size_of_total_reqs` out of sync with the request array in https://github.com/ICLDisco/parsec/pull/521/files#diff-0d6a2a76e0f5fedd623964083223b7f29041101eade63cb104701bb66750d761R921

Reply:
`current_size_of_total_reqs` is the current size of the arrays while `size_of_total_reqs` is the size it should be.

Reply:
The all the assert and checks for posting in the static array (https://github.com/ICLDisco/parsec/pull/521/files#diff-0d6a2a76e0f5fedd623964083223b7f29041101eade63cb104701bb66750d761R331) should use `current_size_of_total_reqs`?

Reply:
good catch !